server/map_tree_updater.go

```diff
@@ -168,20 +159,49 @@ func (t *mapTreeUpdater) newTXFunc(tx storage.MapTreeTX) txFunc {
 	}
 }
 
-// doPreload causes the subtreeCache in tx to become populated with all
-// subtrees on the Merkle path for the indices specified in upd.
-// This is a performance workaround for locking issues which occur when the
-// sparse Merkle tree code is used with a single transaction (and therefore a
-// single subtreeCache too).
-func doPreload(ctx context.Context, tx storage.MapTreeTX, depth uint, upd []smt.Node, rev int64) (map[tree.NodeID2][]byte, error) {
+// doPreload loads all Merkle tree tiles that will be involved when updating
+// the given set of nodes at the specified revision.
+func (t *mapTreeUpdater) doPreload(ctx context.Context, tx storage.MapTreeTX, nodes []smt.Node, rev int64) (*smt.TileSet, error) {
 	ctx, spanEnd := spanFor(ctx, "doPreload")
 	defer spanEnd()
-
 	// TODO(pavelkalinnikov): Avoid using HStar3 directly.
-	hs, err := smt.NewHStar3(upd, nil, depth, 0)
+	hs, err := smt.NewHStar3(nodes, nil, uint(t.hasher.BitLen()), 0)
 	if err != nil {
 		return nil, err
 	}
-	acc := txAccessor{tx: tx, rev: rev + 1}
-	return acc.Get(ctx, hs.Prepare())
+	return t.load(ctx, tx, hs.Prepare(), rev)
+}
+
+// load fetches Merkle tree tiles containing the given node IDs.
+func (t *mapTreeUpdater) load(ctx context.Context, tx storage.MapTreeTX, ids []tree.NodeID2, rev int64) (*smt.TileSet, error) {
+	tileIDs := toTileIDs(ids, t.layout)
+	tiles, err := tx.GetTiles(ctx, rev, tileIDs)
+	if err != nil {
+		return nil, err
+	}
+	ts := smt.NewTileSet(t.tree.TreeId, t.hasher, t.layout)
```

Comment:
Is a TileSet supposed to be an immutable object? I'm surprised to see an Add operation

Reply:
There is a TODO for `TileSet` to make it immutable.